src/orange/detail/BIHMaker.cc

```diff
@@ -0,0 +1,329 @@
+        }
+        centers[i] = center;
+    }
+    return centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create sorted and uniquified X, Y, Z values of bbox centers.
+ */
+BIHMaker::AxesCenters BIHMaker::axes_centers(VecReal3 const& centers)
+{
+    CELER_EXPECT(!centers.empty());
+
+    AxesCenters axes_centers{{}, {}, {}};
+
+    for (auto center : centers)
+    {
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            axes_centers[ax].push_back(center[ax]);
+        }
+    }
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        sort_and_uniquify(axes_centers[ax]);
+    }
+
+    return axes_centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Bounding box of a collection of bounding boxes.
+ */
+BoundingBox BIHMaker::meta_bbox(VecIndices const& indices)
+{
+    CELER_EXPECT(!indices.empty());
+
+    auto inf = std::numeric_limits<double>::infinity();
+
+    Real3 lower = {inf, inf, inf};
+    Real3 upper = {-inf, -inf, -inf};
+
+    for (auto id : indices)
+    {
+        auto bbox = bboxes_[id.unchecked_get()];
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            lower[ax] = std::min(lower[ax], bbox.lower()[ax]);
+            upper[ax] = std::max(upper[ax], bbox.upper()[ax]);
+        }
+    }
+
+    return {lower, upper};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a vector of axes sorted from longest to shortest.
+ */
+BIHMaker::VecAxes BIHMaker::sort_axes(BoundingBox const& bbox)
+{
+    VecAxes axes;
+    std::vector<double> lengths;
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        axes.push_back(axis);
+        lengths.push_back(bbox.upper()[ax] - bbox.lower()[ax]);
+    }
+
+    std::sort(axes.begin(), axes.end(), [&](Axis axis1, Axis axis2) {
+        return lengths[to_int(axis1)] > lengths[to_int(axis2)];
+    });
+    return axes;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Check that only the first bounding box (i.e. exterior volume) is fully inf.
+ */
+bool BIHMaker::check_bbox_extents()
+{
+    return this->fully_inf(bboxes_[0])
+           && std::all_of(
+               bboxes_.begin() + 1, bboxes_.end(), [this](BoundingBox bbox) {
+                   return !this->fully_inf(bbox);
+               });
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Check if a bounding box spans (-inf, inf) in every direction.
+ */
+bool BIHMaker::fully_inf(BoundingBox const& bbox)
+{
+    auto max_double = std::numeric_limits<double>::max();
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        if (bbox.lower()[ax] > -max_double || bbox.upper()[ax] < max_double)
```

Comment:
why not `std::isinf(bbox.lower())`?

Reply:
So I wasn't sure how to handle this. Basically in Shift, when `std::numeric_limits<double>::infinity()` is encountered, I write this to .json as `std::numeric_limits<double>::max()`, rather then some up with some ad hoc token. So this line of code is handling both the Celeritas_Converter case with`std::numeric_limits<double>::infinity()` and the .json canse with `std::numeric_limits<double>::max()`. I could also fix this in the Celeritas json reader.

Reply:
Was the thumbs up for leave as is, or fix in the reader?

Reply:
Good reason to leave as is, but I like your suggestion of moving it to the reader (changing max to inf in the BoundingBoxIO.json.cc)

Reply:
This is going to be part of BoundingBoxUtils, so I think does make sense to test for both `max` and `infinity`.